File: ECommerceApp/DetailViewController.m
//
//  DetailViewController.m
//  ECommerceApp
//
//  Created by Prateek Raj on 10/07/18.
//  Copyright © 2018 Prateek Raj. All rights reserved.
//

#import "DetailViewController.h"

@interface DetailViewController () {
    
    Product *productDetail;
    
    Taxes *tax;
    
    NSArray *variantArray;
}

@property (weak, nonatomic) IBOutlet UILabel *productLabel;
@property (weak, nonatomic) IBOutlet UILabel *dateLabel;

@property (weak, nonatomic) IBOutlet UILabel *shareLabel;
@property (weak, nonatomic) IBOutlet UILabel *viewLabel;
@property (weak, nonatomic) IBOutlet UILabel *orderLabel;

@property (weak, nonatomic) IBOutlet UITableView *pricesTableView;

@end

@implementation DetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void)viewDidAppear:(BOOL)animated {
    
    [super viewDidAppear:animated];
    
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    
    formatter.dateFormat = @"EEEE, MMM d, yyyy";
    
    [self.productLabel setText:productDetail.name];
    [self.dateLabel setText:[formatter stringFromDate:productDetail.dateAdded]];
    
    [self.shareLabel setText:[NSString stringWithFormat:@"Shares:%d",productDetail.shareCount]];
    [self.viewLabel setText:[NSString stringWithFormat:@"Views:%d",productDetail.viewCount]];
    [self.orderLabel setText:[NSString stringWithFormat:@"Orders:%d",productDetail.orderCount]];
    
    [self.pricesTableView reloadData];
}

- (void)loadProductDetails:(Product *)product {
    
    tax = product.taxes;
    
    variantArray = product.variants.allObjects;
    
    productDetail = product;
}

#pragma mark - Table View DataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return variantArray.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"PRICESCELL"];
    
    if (!cell)
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"PRICESCELL"];
    
    UILabel *colorLabel = [cell.contentView viewWithTag:1];
    UILabel *sizeLabel = [cell.contentView viewWithTag:2];
    UILabel *priceLabel = [cell.contentView viewWithTag:3];
    UILabel *taxlabel = [cell.contentView viewWithTag:4];
    UILabel *totalLabel = [cell.contentView viewWithTag:5];
    
    Variant *variant = variantArray[indexPath.row];
    
    [colorLabel setText:variant.color];
    [sizeLabel setText:[NSString stringWithFormat:@"%d",variant.size]];
    [priceLabel setText:[NSString stringWithFormat:@"%d",variant.price]];
    [taxlabel setText:[NSString stringWithFormat:@"%.02f",tax.value]];
    [totalLabel setText:[NSString stringWithFormat:@"%.02f",(variant.price*tax.value)]];
    
    return cell;
}

@end

File: ECommerceApp/DetailViewController.swift
//
//  DetailViewController.swift
//  ECommerceApp
//
//  Created by Prateek Raj on 28/07/18.
//  Copyright © 2018 Prateek Raj. All rights reserved.
//

import UIKit

class DetailViewController: UIViewController {
    fileprivate var productDetail: Product!
    fileprivate var tax: Taxes?
    fileprivate var variantArray: [Variant] = []
    
    @IBOutlet weak var productLabel: UILabel!
    @IBOutlet weak var dateLabel: UILabel!
    @IBOutlet weak var shareLabel: UILabel!
    @IBOutlet weak var viewLabel: UILabel!
    @IBOutlet weak var orderLabel: UILabel!
    @IBOutlet weak var pricesTableView: UITableView!
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE, MMM d, yyyy"
        
        productLabel.text = productDetail.name
        dateLabel.text = formatter.string(from: productDetail.dateAdded as Date? ?? Date())
        shareLabel.text = "Shares: \(productDetail.shareCount)"
        viewLabel.text = "Views: \(productDetail.viewCount)"
        orderLabel.text = "Orders: \(productDetail.orderCount)"
        pricesTableView.reloadData()
    }
    
    func loadProductDetails(_ product: Product) {
        productDetail = product
        tax = product.taxes
        variantArray = product.variants?.allObjects as? [Variant] ?? []
    }
}

extension DetailViewController: UITableViewDataSource, UITableViewDelegate {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return variantArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell =  tableView.dequeueReusableCell(withIdentifier: "PRICESCELL") ?? UITableViewCell(style: .default, reuseIdentifier: "PRICESCELL")
        let colorLabel = cell.contentView.viewWithTag(1) as? UILabel
        let sizeLabel = cell.contentView.viewWithTag(2) as? UILabel
        let priceLabel = cell.contentView.viewWithTag(3) as? UILabel
        let taxlabel = cell.contentView.viewWithTag(4) as? UILabel
        let totalLabel = cell.contentView.viewWithTag(5) as? UILabel
        
        let variant = variantArray[indexPath.row]
        colorLabel?.text = variant.color
        sizeLabel?.text = "\(variant.size)"
        priceLabel?.text = "\(variant.price)"
        taxlabel?.text = String(format: "%.02f", tax?.value ?? 0)
        totalLabel?.text = String(format: "%.02f", (Float(variant.price) + (((tax?.value ?? 0)*Float(variant.price))/100)))
        
        return cell
    }
}
